swarm/api/http/server.go

```diff
@@ -290,6 +291,111 @@ func (s *Server) HandleDelete(w http.ResponseWriter, r *Request) {
 	fmt.Fprint(w, newKey)
 }
 
+func (s *Server) HandlePostDb(w http.ResponseWriter, r *Request) {
+	if r.ContentLength == 0 {
+		frequency, err := strconv.ParseUint(r.uri.Path, 10, 64)
+		if err != nil {
+			w.WriteHeader(http.StatusBadRequest)
+			http.ServeContent(w, &r.Request, "", time.Now(), bytes.NewReader([]byte(err.Error())))
+			return
+		}
+		err = s.api.DbCreate(r.uri.Addr, frequency)
+		if err != nil {
+			w.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+	} else {
+		data, err := ioutil.ReadAll(r.Body)
+		if err != nil {
+			w.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+		_, _, _, err = s.api.DbUpdate(r.uri.Addr, data)
+		if err != nil {
+			w.WriteHeader(http.StatusUnauthorized)
+			http.ServeContent(w, &r.Request, "", time.Now(), bytes.NewReader([]byte(err.Error())))
+			return
+		}
+	}
+	w.WriteHeader(http.StatusOK)
+}
+
+// Retrieve mutable resource updates:
+// bzz-db[-[immutable|-raw]]://<id> - get latest update
+// bzz-db[-[immutable|-raw]]://<id>/<n> - get latest update on period n
+// bzz-db[-[immutable|-raw]]://<id>/<n>/<m> - get update version m of period n
+func (s *Server) HandleGetDb(w http.ResponseWriter, r *Request) {
+	w.Header().Set("Content-Type", "application/octet-stream")
+
+	rootKey, err := s.api.Resolve(r.uri)
+	if err != nil {
+		s.Error(w, r, fmt.Errorf("error resolving %s: %s", r.uri.Addr, err))
+		return
+	}
+
+	var params []string
+	if len(r.uri.Path) > 0 {
+		params = strings.Split(r.uri.Path, "/")
+	}
+	var updateKey storage.Key
+	var period uint64
+	var version uint64
+	var data io.ReadSeeker
+	var dataLength int
+	now := time.Now()
+	switch len(params) {
+	case 0:
+		updateKey, data, dataLength, err = s.api.DbLookup(rootKey, r.uri.Addr, 0, 0)
+		break
+	case 2:
+		version, err = strconv.ParseUint(params[1], 10, 32)
+		if err != nil {
+			break
+		}
+	case 1:
+		period, err = strconv.ParseUint(params[0], 10, 32)
+		if err != nil {
+			break
+		}
+		updateKey, data, dataLength, err = s.api.DbLookup(rootKey, r.uri.Addr, uint32(period), uint32(version))
+		break
+	default:
+		w.WriteHeader(http.StatusBadRequest)
+		err = fmt.Errorf("params 0-2")
+	}
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+	if !r.uri.DbRaw() {
+		entry := api.ManifestEntry{
+			Hash:        rootKey.Hex(),
+			Path:        updateKey.Hex(),
+			ContentType: api.DbManifestType,
+			Size:        int64(dataLength),
+			ModTime:     now,
+			Status:      http.StatusOK,
+		}
+		mode := (6 << 2) | (4 << 1) | 4
```

Comment:
This looks like magic, what are all these numbers for? Please add a comment :wink:.

Reply:
"file mode" 644, or 664 (actually should be 666) if validator isn't set.

Reply:
Ah ok, so you should probably just use an octal literal like `0644`.

Note that what you have here is _not_ the same as octal `0644`, you would instead want:

```
(6 << 6) | (4 << 3) | 4
```

Reply:
Indeed it is.